clients/src/test/java/org/apache/kafka/common/requests/RequestResponseTest.java

```diff
@@ -179,6 +183,34 @@ public void testRequestHeaderWithNullClientId() {
         assertEquals("", deserialized.clientId()); // null is defaulted to ""
     }
 
+    @Test
+    public void testNullMetadataInOffsetCommitRequest() {
+        TopicPartition tp = new TopicPartition("test", 0);
+
+        Map<TopicPartition, OffsetCommitRequest.PartitionData> commitData = new HashMap<>();
+        commitData.put(tp, new OffsetCommitRequest.PartitionData(100, null));
+
+        Map<Integer, OffsetCommitRequest> requests = new HashMap<>();
+        requests.put(0, new OffsetCommitRequest("groupId", commitData));
+        requests.put(1, new OffsetCommitRequest("groupId", 0, "memberId", commitData));
+        requests.put(2, new OffsetCommitRequest("groupId", 0, "memberId", 100000, commitData));
+
+        for (Map.Entry<Integer, OffsetCommitRequest> requestEntry : requests.entrySet()) {
+            Integer version = requestEntry.getKey();
+            OffsetCommitRequest request = requestEntry.getValue();
+
+            // write the raw struct which won't override the null with the default
+            ByteBuffer buffer = ByteBuffer.allocate(request.struct.sizeOfRaw());
+            request.struct.writeTo(buffer);
```

Comment:
Yeah, this was my reluctance to pull the "raw" notion to top level. After talking with @ewencp, it seems like it may be worth it in the long term.